components/brave_federated_learning/brave_operational_profiling.cc

```diff
@@ -0,0 +1,201 @@
+
+  LoadPrefs();
+  reuseOrRefreshEphemeralId();
+}
+
+void BraveOperationalProfiling::LoadPrefs() {
+  platform_ = GetPlatformIdentifier();
+  last_checked_slot_ = pref_service_->GetInteger(kLastCheckedSlotPrefName);
+  ephemeral_id_ = pref_service_->GetString(kEphemeralIDPrefName);
+  ephemeral_id_expiration_ =
+      pref_service_->GetTime(kEphemeralIDExpirationPrefName);
+}
+
+void BraveOperationalProfiling::SavePrefs() {
+  pref_service_->SetInteger(kLastCheckedSlotPrefName, last_checked_slot_);
+  pref_service_->SetString(kEphemeralIDPrefName, ephemeral_id_);
+  pref_service_->SetTime(kEphemeralIDExpirationPrefName,
+                         ephemeral_id_expiration_);
+}
+
+void BraveOperationalProfiling::OnCollectionSlotStartTimerFired() {
+  simulate_local_training_step_timer_->Reset();
+}
+
+void BraveOperationalProfiling::OnSimulateLocalTrainingStepTimerFired() {
+  SendCollectionSlot();
+}
+
+void BraveOperationalProfiling::SendCollectionSlot() {
+  current_collected_slot_ = GetCurrentCollectionSlot();
+  if (current_collected_slot_ == last_checked_slot_) {
+    return;
+  }
+
+  reuseOrRefreshEphemeralId();
+
+  auto resource_request = std::make_unique<network::ResourceRequest>();
+  resource_request->url = operational_profiling_endpoint_;
+  resource_request->headers.SetHeader("X-Brave-Trace", "?1");
+
+  resource_request->credentials_mode = network::mojom::CredentialsMode::kOmit;
+  resource_request->method = "POST";
+
+  url_loader_ = network::SimpleURLLoader::Create(
+      std::move(resource_request), GetNetworkTrafficAnnotationTag());
+  url_loader_->AttachStringForUpload(BuildTraceCollectionPayload(),
+                                     "application/base64");
+
+  url_loader_->DownloadToStringOfUnboundedSizeUntilCrashAndDie(
+      url_loader_factory_.get(),
+      base::BindOnce(&BraveOperationalProfiling::OnUploadComplete,
+                     base::Unretained(this)));
+}
+
+void BraveOperationalProfiling::OnUploadComplete(
+    std::unique_ptr<std::string> response_body) {
+  if (url_loader_->ResponseInfo()->headers->response_code() == 200) {
+    last_checked_slot_ = current_collected_slot_;
+    SavePrefs();
+  }
+}
+
+std::string BraveOperationalProfiling::BuildTraceCollectionPayload() const {
+  base::Value root(base::Value::Type::DICTIONARY);
+
+  root.SetKey("ephemeral_id", base::Value(ephemeral_id_));
+  root.SetKey("platform", base::Value(platform_));
+  root.SetKey("collection_slot", base::Value(current_collected_slot_));
+
+  std::string result;
+  base::JSONWriter::Write(root, &result);
+
+  return result;
+}
+
+int BraveOperationalProfiling::GetCurrentCollectionSlot() const {
+  base::Time::Exploded now;
+  base::Time::Now().LocalExplode(&now);
+
+  return (now.hour * 60 + now.minute) / collection_slot_size_;
+}
+
+std::string BraveOperationalProfiling::GetPlatformIdentifier() {
+  return brave_stats::GetPlatformIdentifier();
+}
+
+void BraveOperationalProfiling::reuseOrRefreshEphemeralId() {
```

Comment:
maybe rename according to brave ads convention to something like `MaybeResetEphemeralId()`

Reply:
👍